File: 300.500/448.FindAllNumbersDisappearedInAnArray.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

class FindAllNumbersDisappearedInAnArray
{
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        int n = nums.size();
        for (int i = 0;i < n;++i) {
            int x = abs(nums[i]);
            if (nums[x - 1] > 0) {
                nums[x - 1] *= -1;
            }
        }
        vector<int> ret;
        for (int i = 0;i < n;++i) {
            if (nums[i] > 0)
                ret.emplace_back(i + 1);
        }
        return ret;
    }
};

void Print(vector<int>& x) {
    cout << "[";
    for (int i : x) {
        cout << i << ",";
    }
    cout << "]" << endl;
}

int main(int argc, char const* argv[])
{
    FindAllNumbersDisappearedInAnArray cls;
    auto& ret = cls.findDisappearedNumbers(vector<int>{4,3,2,7,8,2,3,1});
    Print(ret);
    auto& ret1 = cls.findDisappearedNumbers(vector<int>{1,1});
    Print(ret1);
    return 0;
}

File: go/300.500/448.FindAllNumbersDisappearedInAnArray.go
package main

import "fmt"

func findDisappearedNumbers(nums []int) []int {
	for i := range nums {
		x := abs(nums[i])
		if nums[x-1] > 0 {
			nums[x-1] *= -1
		}
	}
	ret := []int{}
	for i, x := range nums {
		if x > 0 {
			ret = append(ret, i+1)
		}
	}
	return ret
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	fmt.Println(findDisappearedNumbers([]int{4, 3, 2, 7, 8, 2, 3, 1}))
	fmt.Println(findDisappearedNumbers([]int{1, 1}))
}
